worlds/kh2/Rules.py

```diff
@@ -1,96 +1,1839 @@
+        return state.has(ItemName.ThunderElement, self.player, 3)
+
+    def hundred_acre_unlocked(self, state: CollectionState, amount) -> bool:
+        return state.has(ItemName.TornPages, self.player, amount)
+
+    def final_form_region_access(self, state: CollectionState) -> bool:
+        """
+        Can reach one of TT3,Twtnw post Roxas, BC2, LoD2 or PR2
+        """
+        # tt3 start, can beat roxas, can beat gr2, can beat xaldin, can beat storm rider.
+        final_leveling_access = [LocationName.MemorysSkyscaperMythrilCrystal, LocationName.GrimReaper2,
+                                 LocationName.Xaldin, LocationName.StormRider, LocationName.SunsetTerraceAbilityRing]
+        return any(
+                self.world.multiworld.get_location(location, self.player).can_reach(state) for location in
+                final_leveling_access
+        )
+
+    def limit_form_region_access(self, state: CollectionState) -> bool:
+        """
+        multi_form_region_access + namine sketches for leveling
+        """
+        multi_form_region_access = {
+            ItemName.CastleKey,
+            ItemName.BattlefieldsofWar,
+            ItemName.SwordoftheAncestor,
+            ItemName.BeastsClaw,
+            ItemName.BoneFist,
+            ItemName.SkillandCrossbones,
+            ItemName.Scimitar,
+            ItemName.MembershipCard,
+            ItemName.IceCream,
+            ItemName.WaytotheDawn,
+            ItemName.IdentityDisk,
+            ItemName.NamineSketches
+        }
+        return state.has_any(multi_form_region_access, self.player)
+
+    def multi_form_region_access(self, state: CollectionState) -> bool:
+        """
+        Valor, Wisdom and Master Form region access.
+        Note: This does not account for having the drive form. See drive_form_unlock
+        """
+        # todo: if boss enemy start the player with oc stone because of cerb
+        multi_form_region_access = {
+            ItemName.CastleKey,
+            ItemName.BattlefieldsofWar,
+            ItemName.SwordoftheAncestor,
+            ItemName.BeastsClaw,
+            ItemName.BoneFist,
+            ItemName.SkillandCrossbones,
+            ItemName.Scimitar,
+            ItemName.MembershipCard,
+            ItemName.IceCream,
+            ItemName.WaytotheDawn,
+            ItemName.IdentityDisk,
+        }
+        return state.has_any(multi_form_region_access, self.player)
+
+    def level_locking_unlock(self, state: CollectionState, amount):
+        return amount <= sum([state.count(item_name, self.player) for item_name in visit_locking_dict["2VisitLocking"]])
+
+    def summon_levels_unlocked(self, state: CollectionState, amount) -> bool:
+        return amount <= sum([state.count(item_name, self.player) for item_name in summons])
+
+    def set_kh2_rules(self) -> None:
+        world = self.world.multiworld
+        player = self.player
+        for region in world.get_regions(self.player):
```

Comment:
this should be `for region_name, rule in self.region_rules.items():` the reason i grab all my players' regions in messenger is because i also loop through the locations and some of the locations that get their rules set there are in regions that don't have rules